File: ext/java/devs/CoupledModel.java
package devs;

import org.jruby.Ruby;
import org.jruby.RubyArray;
import org.jruby.RubyClass;
import org.jruby.RubyObject;
import org.jruby.anno.JRubyClass;

import devs.Model;
import org.jruby.anno.JRubyMethod;
import org.jruby.runtime.Block;
import org.jruby.runtime.ThreadContext;
import org.jruby.runtime.Visibility;
import org.jruby.runtime.builtin.IRubyObject;

@JRubyClass(name = "DEVS::CoupledModel")
public class CoupledModel extends Model {
    public CoupledModel(Ruby rb, RubyClass klass) {
        super(rb, klass);
    }

    @JRubyMethod(name = "each_coupling", visibility = Visibility.PRIVATE)
    public IRubyObject eachCoupling(ThreadContext context, IRubyObject ary, IRubyObject port, Block block) {
        Ruby rb = getRuntime();
        RubyArray all = (RubyArray) ary;
        RubyArray couplings;

        if (port.isNil()) {
            couplings = all;
        } else {
            couplings = RubyArray.newArray(rb);

            for (int i = 0; i < all.getLength(); i++) {
                RubyObject coupling = (RubyObject) all.entry(i);
                IRubyObject portSrc = coupling.getInstanceVariable("@port_source");

                if (portSrc == port) {
                    couplings.append(coupling);
                }

                if (block.isGiven()) {
                    block.yield(context, coupling);
                }
            }
        }

        return couplings;
    }
}

File: ext/devs/coupled_model.c
#include <coupled_model.h>

VALUE cDEVSCoupledModel;

static VALUE each_coupling(VALUE self, VALUE ary, VALUE port);

/*
* Document-class: DEVS::CoupledModel
*/
void
init_devs_coupled_model() {
    VALUE mdl = rb_define_class_under(mDEVS, "Model", rb_cObject);
    VALUE klass = rb_define_class_under(mDEVS, "CoupledModel", mdl);

    cDEVSCoupledModel = klass;

    rb_define_private_method(klass, "each_coupling", each_coupling, 2);
}


static VALUE
each_coupling(VALUE self, VALUE ary, VALUE port) {
    VALUE couplings;
    long i;

    if (NIL_P(port)) {
        couplings = ary;
    } else {
        couplings = rb_ary_new();

        for (i = 0; i < RARRAY_LEN(ary); i++) {
            VALUE coupling = rb_ary_entry(ary, i);
            VALUE port_src = rb_iv_get(coupling, "@port_source");

            if (port_src == port) {
                rb_ary_push(couplings, coupling);
                if (rb_block_given_p()) {
                    rb_yield(coupling);
                }
            }
        }
    }

    return couplings;
}
